ruby/app.rb

```diff
@@ -601,6 +650,115 @@ def nil_if_empty_envvar(field)
+    pretty_print_response(get_response.to_hash)
+
+    pdf_response = client.cra_check_report_pdf_get(
+      Plaid::CraCheckReportPDFGetRequest.new({ user_token: user_token })
+    )
+
+    content_type :json
+    {
+      report: get_response.report.to_hash,
+      pdf: Base64.encode64(File.read(pdf_response))
+    }.to_json
+  rescue Plaid::ApiError => e
+    error_response = format_error(e)
+    pretty_print_response(error_response)
+    content_type :json
+    error_response.to_json
+  end
+end
+
+def get_cra_base_report_with_retries(plaid_client, user_token)
+  poll_with_retries do
+    plaid_client.cra_check_report_base_report_get(
+      Plaid::CraCheckReportBaseReportGetRequest.new({ user_token: user_token })
+    )
+  end
+end
+
+# Retrieve CRA Income Insights and PDF with Insights
+# Income insights: https://plaid.com/docs/check/api/#cracheck_reportincome_insightsget
+# PDF w/ income insights: https://plaid.com/docs/check/api/#cracheck_reportpdfget
+get '/api/cra/get_income_insights' do
+  begin
+    get_response = get_income_insights_with_retries(client, user_token)
+    pretty_print_response(get_response.to_hash)
+
+    pdf_response = client.cra_check_report_pdf_get(
+      Plaid::CraCheckReportPDFGetRequest.new({ user_token: user_token, add_ons: [Plaid::CraPDFAddOns::CRA_INCOME_INSIGHTS] })
+    )
+
+    content_type :json
+    {
+      report: get_response.report.to_hash,
+      pdf: Base64.encode64(File.read(pdf_response))
+    }.to_json
+  rescue Plaid::ApiError => e
+    error_response = format_error(e)
+    pretty_print_response(error_response)
+    content_type :json
+    error_response.to_json
+  end
+end
+
+def get_income_insights_with_retries(plaid_client, user_token)
+  poll_with_retries do
+    plaid_client.cra_check_report_income_insights_get(
+      Plaid::CraCheckReportIncomeInsightsGetRequest.new({ user_token: user_token })
+    )
+  end
+end
+
+# Retrieve CRA Partner Insights
+# https://plaid.com/docs/check/api/#cracheck_reportpartner_insightsget
+get '/api/cra/get_partner_insights' do
+  begin
+    response = get_check_partner_insights_with_retries(client, user_token)
+    pretty_print_response(response.to_hash)
+
+    content_type :json
+    response.to_hash.to_json
+  rescue Plaid::ApiError => e
+    error_response = format_error(e)
+    pretty_print_response(error_response)
+    content_type :json
+    error_response.to_json
+  end
+end
+
+def get_check_partner_insights_with_retries(plaid_client, user_token)
+  poll_with_retries do
+    plaid_client.cra_check_report_partner_insights_get(
+      Plaid::CraCheckReportPartnerInsightsGetRequest.new({ user_token: user_token })
+    )
+  end
+end
+
+# Since this quickstart does not support webhooks, this function can be used to poll
+# an API that would otherwise be triggered by a webhook.
+# For a webhook example, see
+# https://github.com/plaid/tutorial-resources or
+# https://github.com/plaid/pattern
+def poll_with_retries(ms = 1000, retries_left = 20)
```

Comment:
at some point we should rewrite transactions to use this, but it's not really your problem to do it :-)